### PR TITLE
Improve Bounded Range Expansion

### PR DESCRIPTION
**Pull request by apmoriarty**
- replace most of the bounded index lookup iterator stack with a single BoundedRangeExpansionIterator
- add integration test for bounded index lookup
- add unit test for bounded range expansion iterator
- Integrate tests from commit 7092ee4f877d8931d7368238ad3ca472159ce07a

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/core/iterators/BoundedRangeExpansionIterator.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+        super.init(source, options, env);
+    }
+
+    @Override
+    public IteratorOptions describeOptions() {
+        IteratorOptions opts = new IteratorOptions(getClass().getName(), "Expands bounded ranges using the global index", null, null);
+        opts.addNamedOption(START_DATE, "The start date");
+        opts.addNamedOption(END_DATE, "The end date");
+        opts.addNamedOption(DATATYPES_OPT, "The set of datatypes used to filter keys");
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/core/iterators/BoundedRangeExpansionIterator.java]: Denote that the datatype option is optional

File: warehouse/query-core/src/main/java/datawave/core/iterators/BoundedRangeExpansionIterator.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        }
+
+        String datatype = cq.substring(index + 1);
+        if (!datatypes.isEmpty() && !datatypes.contains(datatype)) {
+            log.trace("datatype {} was filtered out, advancing to next key", datatype);
+            return new FilterResult(false, AdvanceResult.NEXT);
+        }
+
+        if (prevRow != null && prevRow.equals(k.getRow())) {
+            // should never get here
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/core/iterators/BoundedRangeExpansionIterator.java]: I expect we would get here.  Is this not how we advance to the next value?

File: warehouse/query-core/src/main/java/datawave/query/jexl/lookups/BoundedRangeIndexLookup.java [hunk 3]
```diff
@@ -137,20 +134,17 @@ public synchronized void submit() {
...
+                setting.addOption(BoundedRangeExpansionIterator.START_DATE, startDay);
+                setting.addOption(BoundedRangeExpansionIterator.END_DATE, endDay);
+                if (!config.getDatatypeFilter().isEmpty()) {
+                    setting.addOption(BoundedRangeExpansionIterator.DATATYPES_OPT, Joiner.on(',').join(config.getDatatypeFilter()));
+                }
+                bs.addScanIterator(setting);
 
                 // If this is a composite field, with multiple terms, we need to setup our query to filter based on each component of the composite range
-                if (config.getCompositeToFieldMap().get(literalRange.getFieldName()) != null) {
+                if (!config.getCompositeToFieldMap().get(literalRange.getFieldName()).isEmpty()) {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/lookups/BoundedRangeIndexLookup.java:147]: Do we need to check for null as well?

File: warehouse/query-core/src/main/java/datawave/query/jexl/lookups/BoundedRangeIndexLookup.java [hunk 4]
```diff
@@ -239,68 +226,30 @@ protected Callable<Boolean> createTimedCallable(final Iterator<Entry<Key,Value>>
...
-                        String field = null;
+                        // safety check...
+                        Preconditions.checkState(field.equals(literalRange.getFieldName()),
+                                        "Got an unexpected field name when expanding range" + field + " " + literalRange.getFieldName());
 
-                        boolean foundDataType = false;
+                        // obtaining the size of a map can be expensive,
+                        // instead
+                        // track the count of each unique item added.
+                        indexLookupMap.put(field, uniqueTerm);
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/lookups/BoundedRangeIndexLookup.java:247]: So much better.... nice

File: warehouse/query-core/src/main/java/datawave/query/jexl/lookups/BoundedRangeIndexLookup.java (hunk 3 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/jexl/lookups/BoundedRangeIndexLookup.java:147]: We do not. The `Multimap.get()` call returns an empty collection when nothing maps to the provided key. Which means that yes, we were slapping this iterator onto the stack every single time for no reason.

File: warehouse/query-core/src/main/java/datawave/core/iterators/BoundedRangeExpansionIterator.java (hunk 2 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/core/iterators/BoundedRangeExpansionIterator.java]: Comment updated in the next commit. The iterator should only return a single key per unique row in the index, this is more of a safety check than anything else. At one point I had this code block throw an exception and the build passed without issue.

File: warehouse/query-core/src/main/java/datawave/core/iterators/BoundedRangeExpansionIterator.java (hunk 1 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/core/iterators/BoundedRangeExpansionIterator.java]: done

**Review by ivakegg — APPROVED** [review]

**Pull request merged**